connectors/byoc.py

```diff
@@ -178,96 +178,138 @@ async def all_connectors(self):
             yield connector
 
 
-class SyncJob:
-    def __init__(self, elastic_index, connector_id, doc_source=None):
-        self.connector_id = connector_id
-        self.elastic_index = elastic_index
-        self.created_at = datetime.now(timezone.utc)
-        self.completed_at = None
-        self.filtering = Filter()
-        self.client = elastic_index.client
+class ESDocument:
+    def __init__(self, elastic_index, doc_source=None):
+        self.index = elastic_index
         if doc_source is None:
             doc_source = dict()
+        self.id = doc_source.get("_id")
+        self._source = doc_source.get("_source", {})
 
-        self.doc_source = doc_source
-        self.job_id = self.doc_source.get("_id")
-        self.status = str2e(self.doc_source.get("_source", {}).get("status"), JobStatus)
+    def get(self, *keys, default=None):
+        value = self._source
+        for key in keys:
+            if not isinstance(value, dict):
+                return default
+            value = value.get(key)
+        if value is None:
+            return default
+        return value
+
+    async def reload(self):
+        return await self.index.fetch_by_id(self.id)
+
+
+class SyncJob(ESDocument):
+    @property
+    def status(self):
+        return str2e(self.get("status"), JobStatus)
```

Comment:
The default value needs to be an instance of `JobStatus(1)` or `JobStatus.PENDING` etc..

```python
>>> from connectors.byoc import JobStatus
>>> JobStatus
<enum 'JobStatus'>
>>> JobStatus(1)
<JobStatus.PENDING: 1>
```

Reply:
oh sorry, I just realized how the  `str2e` works by grabbing the `Enum` class.

Forget about this comment. On a side note, this is very hard to understand and not semantically correct. We should make this simpler later.

I think we'll want this to work eventually:

```
@property
def status(self):
    return JobStatus(self.get("status"))
```

Reply:
Yes, we discussed this before, this will work:
```python
class JobStatus(Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    CANCELING = "canceling"
    CANCELED = "canceled"
    SUSPENDED = "suspended"
    COMPLETED = "completed"
    ERROR = "error"

# enum to str
JobStatus.PENDING.value

# str to enum
JobStatus("pending")
```

But we also need to consider the edge case when it's `None`, probably as you suggested before, we could have a special `unset` value for each enum.

Reply:
Consolidated the proposal in the comment: https://github.com/elastic/connectors-python/pull/464#issuecomment-1429286586